### PR TITLE
feat: send all flow versions during start

### PR DESCRIPTION
**Pull request by talaharoni**
## Related Issues

Related to https://github.com/descope/etc/issues/329

## Description
On flow start send versions of all flows to be stored on the context.

## Must

- [x] Tests
- [ ] Documentation (if applicable)

### CODE REVIEW

File: packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts [hunk 1]
```diff
@@ -226,6 +226,15 @@ class DescopeWc extends BaseDescopeWc {
     const loginId = this.sdk.getLastUserLoginId();
     const flowConfig = await this.getFlowConfig();
     const projectConfig = await this.getProjectConfig();
+    const flowVersions = {
+      version: flowConfig.version,
+      componentsVersion: projectConfig.componentsVersion,
+      versions: Object.keys(projectConfig.flows || []).reduce((acc, key) => {
```
Comment (asafshen) [inline packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts:232]: why do we need other flows version here?

File: packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts [hunk 2]
```diff
@@ -226,6 +226,15 @@ class DescopeWc extends BaseDescopeWc {
...
+        const flowConf = projectConfig.flows[key];
```
Comment (asafshen) [inline packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts:233]: not critical at all, but I lov to use `ObjectEntries([key, value] => {}` that already get the matching value 
versions: Object.values(projectConfig.flows || []).reduce((acc, [key, flowConf) => {
...}, {})

File: packages/sdks/core-js-sdk/src/sdk/flow/types.ts [hunk 3]
```diff
@@ -6,3 +6,9 @@ type JSONSerializable =
   | Array<JSONSerializable>;
 
 export type FlowInput = Record<string, JSONSerializable>;
+
+export type FlowVersions = {
+  version: number;
+  componentsVersion: string;
+  versions: Record<string, number>;
```
Comment (asafshen) [inline packages/sdks/core-js-sdk/src/sdk/flow/types.ts:13]: maybe worth to name the attributes here `currentVersion` and `otherVersions`? your choice

**Review by asafshen — DISMISSED** [review]
looks good! minor comments / question

File: packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts (hunk 1 above)
Comment (talaharoni) [reply packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts:232]: When starting flow with subflows. we need to know the version for all participate flows.

Comment (asafshen) [reply packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts:232]: cool, maybe we should add something like `// pass other flow versions, that may be used as a part of the current flow`

**Review by asafshen — DISMISSED** [review]

File: packages/sdks/core-js-sdk/src/sdk/flow/index.ts [hunk 4]
```diff
@@ -31,6 +32,7 @@ const withFlow = (httpClient: HttpClient) => ({
           interactionId,
           version,
           componentsVersion,
```
Comment (orius123) [inline packages/sdks/core-js-sdk/src/sdk/flow/index.ts:34]: Why do we send it twice?

File: packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts [hunk 5]
```diff
@@ -508,6 +520,7 @@ class DescopeWc extends BaseDescopeWc {
           interactionId,
           version,
```
Comment (nirgur) [inline packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts:521]: Do we need version as well?

File: packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts [hunk 6]
```diff
@@ -282,6 +293,7 @@ class DescopeWc extends BaseDescopeWc {
           '',
           flowConfig.version,
```
Comment (nirgur) [inline packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts:294]: Do we need to send it?

File: packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts [hunk 7]
```diff
@@ -226,6 +226,17 @@ class DescopeWc extends BaseDescopeWc {
     const loginId = this.sdk.getLastUserLoginId();
     const flowConfig = await this.getFlowConfig();
     const projectConfig = await this.getProjectConfig();
+    const flowVersions = {
```
Comment (nirgur) [inline packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts:229]: I would expect this to be in the BaseDescopeWc class

In addition, why do we need here also?
```
version: flowConfig.version,
componentsVersion: projectConfig.componentsVersion,
```

`componentsVersion` is sent as a separate param, 
and flow version should be already included in `allVersions`

**Review by nirgur — COMMENTED** [review]
👍 
Added a few comments

File: packages/sdks/core-js-sdk/src/sdk/flow/index.ts (hunk 4 above)
Comment (talaharoni) [reply packages/sdks/core-js-sdk/src/sdk/flow/index.ts:34]: I wanted to keep it for backward compatibility so i can remove it in phases. 
But perhaps better to deliver the backend first and only after it's in production to deliver the sdk

File: packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts (hunk 7 above)
Comment (talaharoni) [reply packages/sdks/web-component/src/lib/descope-wc/DescopeWc.ts:229]: i wanted in next phase to remove `version` and `componentsVersion` so they will be only in this object. 
But decided to remove now. 

I don't know in all places the current flowID so prefer to have the option to know it without that

File: packages/sdks/core-js-sdk/src/sdk/flow/index.ts (hunk 4 above)
Comment (orius123) [reply packages/sdks/core-js-sdk/src/sdk/flow/index.ts:34]: You forgot to remove the flow version `version?: number,`

Comment (talaharoni) [reply packages/sdks/core-js-sdk/src/sdk/flow/index.ts]: removed.

**Comment by orius123** [comment]
@nirgur why did the build fail?

**Review by orius123 — APPROVED** [review]

**Comment by nirgur** [comment]
> @nirgur why did the build fail?

e2e tests are not stable

**Comment by talaharoni** [comment]
> > @nirgur why did the build fail?
> 
> e2e tests are not stable

Any way to resolve it? actually it's pretty stable. fails every time :/

**Pull request merged**